File: FlinkTurbineFaultDiagnosis/src/main/java/com/wang/flink/sink/RedisSink.java
package com.wang.flink.sink;

import com.wang.flink.model.Turbine;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommand;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommandDescription;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisMapper;

/**
 * @author 王一宁
 * @date 2020/1/16 9:40
 */
public class RedisSink implements RedisMapper<Turbine> {

    //创建redis连接
    FlinkJedisPoolConfig conf = new FlinkJedisPoolConfig.Builder().setHost("127.0.0.1").build();

    /**
     * 如何调用
     * @return
     *  summed.addSink(new RedisSink<Turbine>(conf,new RedisSink()));
     *
     */


    //写入方法set 多个键匹配一个vlaue
    @Override
    public RedisCommandDescription getCommandDescription() {
        return new RedisCommandDescription(RedisCommand.HSET,"Turbine_Count");
    }

    //联合的key 风机编号+ 风机类型+ 风机经纬度
    @Override
    public String getKeyFromData(Turbine turbine) {
        return turbine.word + "_" +turbine.province +"_" + turbine.longitude;
    }

    //value String
    @Override
    public String getValueFromData(Turbine turbine) {
        return String.valueOf(turbine.counts);
    }
}

File: FlinkTurbineFaultDiagnosis/src/main/scala/com/wang/turbine/sink/RedisSink.scala
package com.wang.turbine.sink

import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.connectors.redis.RedisSink
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig
import org.apache.flink.streaming.connectors.redis.common.mapper.{RedisCommand, RedisCommandDescription, RedisMapper}

object RedisSink {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    //env.setMaxParallelism(1)

    //1.读取文件
    val inputStream = env.readTextFile("D:\\APP\\IDEA\\workplace\\HelloFlink\\src\\main\\resources\\sensor.txt")
    //2.转换操作
    val dataStream = inputStream.map(data=>{
      val dataArray = data.split(",")//trim去掉空格 有的话
      //SensorReading(dataArray(0).trim,dataArray(1).trim.toLong,dataArray(2).trim.toDouble)
    })

    //3.创建jedis连接
    val conf = new FlinkJedisPoolConfig.Builder()
        .setHost("localhost")
        .setPort(6379)
        .build()

    //3.写入redis
    //dataStream.addSink(new RedisSink(conf,new MyRedisMapper()))

    env.execute("redissink")

  }
}


//class MyRedisMapper() extends RedisMapper[SensorReading]{
//  //保存数据到redis的命令
//  override def getCommandDescription: RedisCommandDescription = {
//    //保存成把传感器id，温度值 hash表
//    new RedisCommandDescription(RedisCommand.HSET,"sensor_temperature")
//  }
//
//  //定义保存到redis的key
//  override def getKeyFromData(t: SensorReading): String = t.id
//
//  //定义保存到redis的value
//  override def getValueFromData(t: SensorReading): String = t.temperature.toString
//}